### PR TITLE
Create edit challenge page

### PR DESCRIPTION
**Pull request by yuqi88**
## Description

Please merge #577 before merging this PR.
Add create-edit-challenge component.
Add form for creating challenges.
Add updateChallenge and addChallenge in course service

## Types of changes

What types of changes does your code introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (only improve the code quality no change in functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] Issue has been created
- [ ] Every file touched is linted
- [ ] Every file touched has tests that cover all the funcitonality with green coverage
- [ ] Documentation is updated (if applicable).

## Issue

Please link the issue here.

## Tests

Please describe in detail what tests you added or changed.

### CODE REVIEW

File: src/app/course/_forms/course-event.form.ts [hunk 1]
```diff
@@ -11,6 +11,17 @@ import {
 import {CourseEvent} from "@app/_models"
 import {TuiDay, TuiDayRange, TuiTime} from "@taiga-ui/cdk"
 
+// export interface challengeFormData {
+//     name: string
+//     type: string
+//     challenge_type_value?: any
+//     max_team_size?: number
+//     startEndDatePicker?: TuiDay
```
Comment (keyvankhademi) [inline src/app/course/_forms/course-event.form.ts:19]: names should match the api.

File: src/app/course/_forms/course-event.form.ts [hunk 2]
```diff
@@ -97,4 +108,71 @@ export class CourseEventForm {
...
+            challengeType: new FormControl('', [Validators.required]),
+            challenge_type_value:  new FormControl(3),// challenge_type_value:  new FormControl(3, [Validators.required]),
+            max_team_size: new FormControl(null, [Validators.required]),
+            startEndDatePicker: new FormControl(
+                new TuiDayRange(TuiDay.currentLocal(), TuiDay.currentLocal().append({day: 7})),
+                [Validators.required]
+            ),
+            startTimePicker: new FormControl(TuiTime.currentLocal(), [Validators.required]),
+            endTimePicker: new FormControl(TuiTime.currentLocal(), [Validators.required]),
+            selectChallenges:  new FormControl(null)// selectChallenges:  new FormControl(null, [Validators.required]),
```
Comment (keyvankhademi) [inline src/app/course/_forms/course-event.form.ts:129]: remove the select challenges.

File: src/app/course/_forms/course-event.form.ts [hunk 3]
```diff
@@ -97,4 +108,71 @@ export class CourseEventForm {
...
+            name: new FormControl(null, [Validators.required]),
+            challengeType: new FormControl('', [Validators.required]),
+            challenge_type_value:  new FormControl(3),// challenge_type_value:  new FormControl(3, [Validators.required]),
+            max_team_size: new FormControl(null, [Validators.required]),
+            startEndDatePicker: new FormControl(
+                new TuiDayRange(TuiDay.currentLocal(), TuiDay.currentLocal().append({day: 7})),
+                [Validators.required]
+            ),
+            startTimePicker: new FormControl(TuiTime.currentLocal(), [Validators.required]),
+            endTimePicker: new FormControl(TuiTime.currentLocal(), [Validators.required]),
```
Comment (keyvankhademi) [inline src/app/course/_forms/course-event.form.ts:128]: add the FormArray similar to the goals

File: src/app/course/_forms/course-event.form.ts [hunk 4]
```diff
@@ -11,6 +11,17 @@ import {
 import {CourseEvent} from "@app/_models"
 import {TuiDay, TuiDayRange, TuiTime} from "@taiga-ui/cdk"
 
+// export interface challengeFormData {
```
Comment (keyvankhademi) [inline src/app/course/_forms/course-event.form.ts:14]: Capital C

File: src/app/course/_forms/course-event.form.ts [hunk 5]
```diff
@@ -97,4 +108,71 @@ export class CourseEventForm {
...
+
+    }
+
+    /**
+     * Returns the formatted form data ready to be sent to the backend
+     * @param formData - the data to be formatted, a FormGroup object
+     * @param courseId - the event's courseId
+     * @param eventId - the event's ID if it already exists
+     */
+    static formatChallengeFormData(formData: FormGroup, courseId: number,eventId: number): CourseEvent{
```
Comment (keyvankhademi) [inline src/app/course/_forms/course-event.form.ts:158]: return format should be `ChallengeFormData`

File: src/app/course/challenge/course-challenge-create-edit/course-challenge-create-edit.component.html [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+            <h3 class="tui-form__header">
+                Questions
+                <button
+
+                    icon="tuiIconPlus"
+                    size="s"
+                    tuiButton
+                    type="button"
+                    class="ml-3"
+                    appearance="secondary"
```
Comment (keyvankhademi) [inline src/app/course/challenge/course-challenge-create-edit/course-challenge-create-edit.component.html:91]: after having the form array add the onclick similar to goals.

File: src/app/course/_forms/course-event.form.ts [hunk 7]
```diff
@@ -97,4 +117,127 @@ export class CourseEventForm {
...
+        })
+    }
+
+    //Need a new relationship with the Event model to relate the questions
+    // static createChallengeQuestionSetFromChallengeForm(challenge:CourseEvent): FormGroup {
+    //     const builder = new FormBuilder()
+    //     return builder.group({
+    //         category: new FormControl('', [Validators.required]),
+    //         difficulty: new FormControl('', [Validators.required]),
+    //         number_of_questions: new FormControl('', [Validators.required])
```
Comment (yuqi88) [inline src/app/course/_forms/course-event.form.ts:203]: I'm not sure how I can get those fields from an event object. It needs backend modification, but not sure where to start for the backend.

File: src/app/course/challenge/course-challenge-create-edit/course-challenge-create-edit.component.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    getChallengeQuestionSetFormControls(): FormControl[] {
+        return this.getChallengeQuestionSets().controls as FormControl[]
+    }
+
+    getFormControl(fc: FormControl, field:string): FormControl {
+        return fc.get(field) as FormControl
+    }
+
+    //TODO: Need support from backend
+    getNumQuestionsLimit(fc: FormControl): number {
```
Comment (yuqi88) [inline src/app/course/challenge/course-challenge-create-edit/course-challenge-create-edit.component.ts:75]: Does it need support from the backend? If so, where should I start?

File: src/app/course/challenge/course-challenge-create-edit/course-challenge-create-edit.component.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    onSubmit(): void {
+        console.log('submitted')
+
+        const challengeData = CourseEventForm.formatChallengeFormData(
+            this.challengeForm,
+            this.courseId
+        )
+        //TODO: what about the questions in each challenge?
+        if (this.eventId) {
+            this.courseEventService.updateChallenge(challengeData, this.eventId).subscribe(() => {
```
Comment (yuqi88) [inline src/app/course/challenge/course-challenge-create-edit/course-challenge-create-edit.component.ts:105]: The onSubmit is only able to update or create challenges, but the questions within the challenges are not being dealt with. Where should I start for this in the backend?

File: src/app/course/challenge/course-challenge-snippet/course-challenge-snippet.component.html [hunk 10]
```diff
@@ -40,6 +53,19 @@
         </tui-accordion>
     </ng-container>
     <ng-template #skeletonTemplate>
+        <div class="flex justify-end toolBar">
+            <button
+                *ngIf="course?.has_create_event_permission"
+                class="tui-skeleton"
```
Comment (yuqi88) [inline src/app/course/challenge/course-challenge-snippet/course-challenge-snippet.component.html:59]: The skeleton added to Add challenge button is not showing when loading. I'm not sure why.

**Review by keyvankhademi — APPROVED** [review]

**Pull request merged**